<html>
<head>
  <title>FragSort</title>
</head>
<body>  

<h1>FragSort</h1>


<h3>Table of Contents</h3>
<ol>
  <li><a href=README.html#introduction>Introduction</a></li>
  <li><a href=README.html#require>Requirements</a></li>
  <li><a href=README.html#structure>Files and Directories</a></li>
  <li><a href=README.html#compiling>Compiling</a></li>
  <li><a href=README.html#running>Running FragSort</a></li>
  <li><a href=README.html#sample>Sample Run</a></li>
  <li><a href=README.html#output>Understanding the Output</a></li>
  <li><a href=README.html#missing>Missing Files</a></li>
  <li><a href=README.html#about>About FragSort</a></li>
  <li><a href=README.html#copyright>Copyright and License</a></li>
</ol>


<h2 id="introduction">Introduction</h2>

<p>FragSort (Fragment Sorting) is a system for sorting next generation sequencing data using either quicksort or radix sort (random permutation is also possible to act as a baseline).  This document accompanies the archive, which includes:</p>

<ul>
<li>source code in C++,</li>
<li>some Perl and bash scripts for testing the software, and</li>
<li>a very small data file for testing.</li>
</ul>

<p>This archive does not include any binaries.</p>

<p>This program is described in the paper:</p>

<blockquote>R. Wan and K. Asai.  Sorting next generation sequencing data improves compression effectiveness. In <em>Proc. 2010 IEEE International Conference on Bioinformatics and Biomedicine (BIBM) - Workshops and Posters</em>, pages 567-572, 2010.</blockquote>

<p>which we refer to as "the paper" throughout this document.  In the paper, the name of the software is never mentioned on purpose.  Instead, the paper directly references the algorithms used (radix sort or quicksort).</p>

<h2 id="require">Requirements</h2>

<table border>
<tr>
  <th>Software</th><th>Version</th><th>Required?</th><th>Web site</th>
</tr>
<tr>
  <td>g++</td><td>4.4.5</td><td>Yes</td><td><a href="http://gcc.gnu.org/">http://gcc.gnu.org/</a></td>
</tr>
<tr>
  <td>CMake</td><td>2.8.2</td><td>Yes</td><td><a href="http://www.cmake.org/">http://www.cmake.org/</a></td>
</tr>
<tr>
  <td>Boost library</td><td>1.42.0</td><td>Yes</td><td><a href="http://www.boost.org/">http://www.boost.org/</a></td>
</tr>
<tr>
  <td>Perl</td><td>5.10.1</td><td>No</td><td><a href="http://www.perl.org/">http://www.perl.org/</a></td>
</tr>
<tr>
  <td>bash</td><td>4.1.5</td><td>No</td><td><a href="http://www.gnu.org/software/bash/">http://www.gnu.org/software/bash/</a></td>
</tr>
<caption>Requirements of FragSort</caption>
</table>

<p>Experiments in the paper using FragSort was executed on Linux systems running Debian 6.0 (squeeze) or CentOS 5.4.  Additional software requirements of FragSort is listed in the table above and represent the software versions on which the experiments were run.  They do not represent the minimum requirements; it is possible that lower versions can be used.</p>

<p>The Boost Library must be both installed and compiled to make use of the <b>program_options</b> library.  Under some Linux distributions, Boost can be installed using its associated package manager (such as "aptitude" for Debian and Ubuntu).  Consult your Linux distribution's documentation for further information.</p>  

<p>Perl and bash are optional because of the testing scripts that are provided with the archive.  Of course, they do not have to be used but are provided as a convenience for users.  If they are used, other Perl modules will be required, such as File::Copy and AppConfig.  All of these modules are available as part of your Linux distribution or in the Comprehensive Perl Archive Network (<a href="http://www.cpan.org/">CPAN</a>).</p>


<h2 id="structure">Files and Directories</h2>

<p>After having unarchived the file, the following directory structure should result:</p>

<ul>
<li>fragsort/</li>
  <ul>
  <li>bin/                 - Empty initially</li>
  <li>data/                - Data directory
    <ul>
      <li>sample.fastq       - Sample data file</li>
    </ul>
  </li>
  <li>doc/                 - Documentation directory
    <ul>
    <li>fdl-1.3.txt        - Copy of GNU FDL license v1.3</li>
    <li>gpl.txt            - Copy of GNU GPL license v3</li>
    <li>AUTHORS            - Authors of FragSort</li>
    <li>ChangeLog          - History of changes to FragSort</li>
    <li>NEWS               - Important NEWS</li>
    <li>README.html        - This file in HTML</li>
    <li>README             - This file as plain text</li>
    <li>VERSION            - Current version number</li>
    </ul>
  </li>  
  <li>scripts/             - Scripts for testing fragsort</li>
  <li>src/                 - C++ source directory  <b>[*]</b></li>
    <ul>
    <li>FindBoost.cmake    - Updated version of FindBoost.cmake to help with compiling.  Will be part of Cmake from 2.8.3 RC1.</li>
    </ul>
  </ul>
</ul>

<p><b>[*]</b> Subdirectories not copied during the "make install" (see next section).</p>



<h2 id="compiling">Compiling</h2>

<p>The FragSort software is written in C++ and has been compiled using v4.4.5 of g++.  The system has been tested on a 64-bit system, but it does not make use of any features from 64-bit architectures.</p>

<p>CMake (at least version 2.8) is used to compile the software and it is recommended that an "out-of-source" build is performed so as not to clutter the original source directories.  We give some brief instructions below on how to do this:</p>

<ol>
<li>Install Boost.  Set the variable BOOST_ROOT to the location of Boost if it has not already been set:

      <blockquote><tt>export BOOST_ROOT=/usr/local/boost_1_42_0/</tt></blockquote>
</li>

<li>Expand the FragSort archive that accompanies this README file in a temporary directory.</li>

<li>Within the expanded directory, create a <tt>build/</tt> subdirectory and then enter it.  Then run

      <blockquote><tt>cmake ..</tt></blockquote>

    By default, this will set up a Makefile to install the program into /usr/local.  Without system administrator access, this would be impossible.  To use another directory, type this:

      <blockquote><tt>cmake .. -DCMAKE_INSTALL_PREFIX=~/tmp</tt></blockquote>

    or whichever directory you prefer.
</li>
<li>Type "make" to compile the C++ source code of FragSort.</li>

<li>If all goes well, type "make install" to install the software.  This copies the important files from the archive (see Section III).  The temporary directory with the expanded archive can now be deleted, unless you would like to see the source of the program.</li>
</ol>


<h2 id="running">Running FragSort</h2>

<p>Running FragSort (as "bin/fragsort") with the <b>--help</b> option flag will print a list of options.</p>

<p>Alternatively, a set of scripts is provided in the <tt>scripts/</tt> directory which will allow users to quickly use the software.  See the next section for an example run.</p> 



<h2 id="sample">Sample Run</h2>

<p>A sample data file "<tt>data/sample.data</tt>" is included in the archive.  The reads in this data file is entirely synthetic with both the sequences and the quality scores being generated randomly using a uniform distribution.  So the compression ratio achieved with this data file should not be taken seriously.</p>

<p>In the <tt>scripts/</tt> directory, several Perl scripts and a single bash script is available for testing.  The bash script is the main script and can be executed as follows:</p>

  <blockquote>cd scripts<br>
  ./process.sh sample</blockquote>

<p>(Note that the <tt>.fastq</tt> extension is not included.)  New FASTQ data files can be placed in the <tt>data/</tt> subdirectory and executed in the same manner.  The <tt>process.sh</tt> script runs the following commands:</p>

<ol>
  <li>Show the size of the input file.</li>
  <li>Run <tt>scripts/collect-stats.pl</tt> to show the size of each component of the file.</li>
  <li>Run <tt>scripts/run-fragsort.pl</tt> command to execute Quicksort (only) and outputing to the <tt>~/tmp/1</tt> subdirectory, similar to the non-indexed-based experiments from the BIBM 2010 paper (see Section VIII).</li>
  <li>Run <tt>scripts/run-fragsort-index.pl</tt> command to execute Quicksort (only) and outputing to the <tt>~/tmp/2</tt> subdirectory, similar to the indexed-based experiments from the paper.</li>
</ol>
  
<p>See <tt>scripts/process.sh</tt> or the Perl scripts themselves if you wish to change the selected parameters.  If the command is issued on <tt>sample.fastq</tt>, then the output will be similar to the output in <tt>doc/sample.fastq.out</tt> .</p>


<h2 id="output">Understanding the Output</h2>

<p>In the output directory (specified using the <b>--outdir</b> option to fragsort), a set of files is created.  When the scripts complete, the temporary files are removed.  The remaining files for both <tt>run-fragsort.pl</tt> and <tt>run-fragsort-index.pl</tt> are given in the following tables:</p>

<table border>
<caption>Output from <tt>run-fragsort.pl</tt></caption>
<tr><th>Filename</th><th>Description</th></tr>
<tr><td>sample-[unique ID]-[machine name]-fastq-{bpc,time}.results</td><td>Original FASTQ file compression</td></tr>
<tr><td>sample-[unique ID]-[machine name]-frags-{bpc,time}.results</td><td>Sequence compression</td></tr>
<tr><td>sample-[unique ID]-[machine name]-halffrags-{bpc,time}.results</td><td>Identifier and sequence compression</td></tr>
<tr><td>sample-[unique ID]-[machine name]-halfqscores-{bpc,time}.results</td><td>Identifier and quality scores compression</td></tr>
<tr><td>sample-[unique ID]-[machine name]-oneid-{bpc,time}.results</td><td>Second identifier removed</td></tr>
<tr><td>sample-[unique ID]-[machine name]-qscores-{bpc,time}.results</td><td>Quality scores compression</td></tr>
</table>

<table border>
<caption>Output from <tt>run-fragsort-index.pl</tt></caption>
<tr><th>Filename</th><th>Description</th></tr>
<tr><td>sample-[unique ID]-[machine name]-bit-{bpc,time}.data</td><td>Index (bit-based) compression</td></tr>
<tr><td>sample-[unique ID]-[machine name]-fastq-time.results</td><td>Original FASTQ file compression</td></tr>
<tr><td>sample-[unique ID]-[machine name]-frags-{bpc,time}.data</td><td>Sequence compression</td></tr>
<tr><td>sample-[unique ID]-[machine name]-ids-{bpc,time}.data</td><td>Identifiers compression</td></tr>
<tr><td>sample-[unique ID]-[machine name]-qscores-{bpc,time}.data</td><td>Quality scores compression</td></tr>
</table>

<p>There are minor differences between the formats of each file, which are all tab-separated.  Basically, for the files reporting on compression effectiveness (bpc = bits per character), the fields are:</p>

<ol>
<li>filename</li>
<li>method (quicksort, radix sort, or random)</li>
<li>Quicksort or radix sort parameter, as a percentage</li>
<li>Replicate number</li>
<li>gzip size</li>
<li>bzip2 size</li>
<li>Size of the input data</li>
<li>Size of the original FASTQ data file</li>
</ol>

<p>To get the compression ratio of gzip, divide the value in the 5th column by the value in either the 7th or 8th column (depending which is the reference point you desire).  For the time files, the user and elapsed times are reported once per line.  See <tt>scripts/help_fragsort.pm</tt> for more information about the format of the output files.</p>


<h2 id="missing">Missing Files</h2>

<p>In the paper, we employ delta coding to encode the generated index.  Delta coding is <b>not</b> included in this archive as the source is being cleaned up.  Hopefully, it will be released in a few months.</p>

<p>Instead, binary coding using a fixed-length of 32 bits is used (i.e., no coding).  As our experiments showed that delta coding was only slightly better than binary coding, this change does not impact our results significantly.</p>



<h2 id="about">About FragSort</h2>

<p>This software was implemented by Raymond Wan:
  <blockquote>
  E-mail:  <a href="r-wan@cb.k.u-tokyo.ac.jp">r-wan@cb.k.u-tokyo.ac.jp</a><br>
  Homepage:  <a href="http://www.cbrc.jp/~rwan/">http://www.cbrc.jp/~rwan/</a>
  </blockquote>
</p>

<p>The latest version of FragSort can be downloaded from the above web page.  This software is actively maintained.  If you have any information about bugs, suggestions for the documentation or just have some general comments, feel free to write to the above address.</p>

<p>The following tools was used for the development of FragSort:</p>

<ul>
<li><a href="http://gcc.gnu.org/">g++ compiler</a></li>
<li><a href="http://www.bash.org/">bash</a></li>
<li><a href="http://www.perl.org/">perl</a></li>
<li><a href="http://www.cmake.org/">CMake</a></li>
<li><a href="http://www.boost.org/">Boost libraries for C++</a></li>
<li><a href="http://www.mbayer.de/html2text/">html2text</a></li>
</ul>


<h2 id="copyright">Copyright and License</h2>

<address>FragSort (Fragment Sorting)<br>
Copyright (C) 2010  by Raymond Wan</address>

<p>FragSort is distributed under the terms of the GNU General Public License (GPL, version 3 or later) -- see the file <tt>doc/gpl.txt</tt> for details.</p>

<p>Permission is granted to copy, distribute and/or modify this document under the terms of the GNU Free Documentation License, Version 1.3 or any later version published by the Free Software Foundation; with no Invariant Sections, no Front-Cover Texts and no Back-Cover Texts.  A copy of the license is included with the archive as <tt>doc/fdl-1.3.txt</tt>.</p>

<p>Document Version:  $Id: README.html 42 2010-12-27 08:24:30Z rwan $</p>

</body>
</html>

